\name{plotGene}
\alias{plotGene}
%- Also NEED an '\alias' for EACH other topic documented here.
\title{
plotGene & plotGenePair
}
\description{
function to plot average coverage of a given gene using mapped bam files
}
\usage{
plotGene(IP_BAM, INPUT_BAM, size.IP, size.INPUT, geneName, geneModel, center = mean ,ZoomIn=NULL)
}
%- maybe also 'usage' for other objects documented here.
\arguments{
  \item{IP_BAM}{ a vector of bam file path for IP sample}
  \item{IUPUT_BAM}{ a vector of bam file path for input sample}
  \item{size.IP}{ a vector of size factor (numbers close to 1) that are proportional to IP sample library sizes}
  \item{size.INPUT}{ a vector of size factor (numbers close to 1) that are proportional to input sample library sizes}
  \item{geneName}{ charactor, the gene name (as defined in gtf file) to be plotted }
  \item{geneModel}{ a Grange List containing gene model. It is made by gtfToGeneModel(gtf)}
  \item{center}{ define the method to center the coverage of multiple replicates. Can be "mean" or "median"}
  \item{ZoomIn}{ A vector of number defining the genomic coordinate range within the given gene to zoom in the plot}

}
\details{
%%  ~~ If necessary, more details than the description above ~~
}
\value{
A plot generated by ggplot2
}
\references{
%% ~put references to the literature/web site here ~
}
\author{
%%  ~~who you are~~
}
\note{
%%  ~~further notes~~
}

%% ~Make other sections like Warning with \section{Warning }{....} ~

\seealso{
%% ~~objects to See Also as \code{\link{help}}, ~~~
}
\examples{
%%
}
% Add one or more standard keywords, see file 'KEYWORDS' in the
% R documentation directory.
\keyword{ ~kwd1 }% use one of  RShowDoc("KEYWORDS")
\keyword{ ~kwd2 }% __ONLY ONE__ keyword per line
